Cache emulated hue states attributes between get and put calls to avoid unexpected alexa errors

jyavenard opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
None

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
When Alexa issues a command to a Hue hub; it immediately queries the hub for the entity's state.to confirm if the command was successful.
It expects the state to be effective immediately.
There may be a delay for the new state to actually be active, this is particularly obvious when using group lights.

This leads Alexa to report that the light had an error.

We now cache the state sets by Alexa and return those cached values for up to two seconds.

Fixes #38446


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->
No changes required anywhere.

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #38446
- This PR is related to issue: #35307
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jyavenard committed:
Wait for the state of the entity to actually change before resolving PUT request

Additionally, we cache the entity's properties for up to two seconds for the successive GET state request

When Alexa issues a command to a Hue hub; it immediately queries the hub for the entity's state to confirm if the command was successful.
It expects the state to be effective immediately after the PUT request has been completed.
There may be a delay for the new state to actually be active, this is particularly obvious when using group lights.
This leads Alexa to report that the light had an error.

So we wait for the state of the entity to actually change before responding to the PUT request.

Due to rounding issue when converting the HA range (0..255) to Hue range (1..254) we now cache the state sets by Alexa and return those cached values for up to two seconds so that Alexa gets the same value as it originally set.

Fixes #38446

jyavenard commented:
I can't make any sense of the codecov failure; it shows changes in files that haven't been modified and in the file modified change of coverage involve lines that haven't changed (just got shifted due to the comment at the top)

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -556,16 +561,25 @@ async def put(self, request, username, entity_number):
+    cached_state = None
+
+    # Check if we have a cached entry, and if so if it hasn't expired.
+    if cached_state_entry is not None:
+        if (
+            cached_state_entry[1] is None
+            or time.time() - cached_state_entry[1] < STATE_CACHED_TIMEOUT
+        ):
+            cached_state = cached_state_entry[0]
```
Please make these constants
```suggestion
    if cached_state_entry is not None:
        if (
            cached_state_entry[1] is None
            or time.time() - cached_state_entry[STATE_CACHE_STATE] < STATE_CACHED_TIMEOUT
        ):
            cached_state = cached_state_entry[STATE_CACHE_TIME]
```

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -520,7 +523,9 @@ async def put(self, request, username, entity_number):
-            config.cached_states[entity_id] = parsed
+            config.cached_states[entity_id] = [parsed, None]
+        else:
+            config.cached_states[entity_id] = [parsed, time.time()]
```
This is going to obscure any failures.  We can call `async_track_state_change_event` and store the callable it returns in the cache as well as an `asyncio.Event`.  Once the state change fires we can set the event.  Then in the `get_entity_state` we can wait for the event for up to the timeout (maybe 2 seconds) and verify the entity state matches the cache (ignoring brightness).  If it matches, return the cache, if it doesn't match return the current state.

jyavenard replied:
I don't think that obscuring any failures is what you want if you want to properly mimic how a Hue hub work.
A real hue device never reports a failure from my testing.

Here is a command with a Hue and a zigbee bulb that has been off for over 10 minutes.
The Hue app will show "All lights are on" or "All lights are off" for the room containing that bulb, even with the light showing up as "unreachable".

Sending the following command over the Hue API will show:
```
$ curl -X PUT -d '{"on":true, "bri":254}' http://192.168.10.180/api/XXX/lights/20/state  && echo && curl http://192.168.10.180/api/XX/lights/20
[{"success":{"/lights/20/state/on":true}},{"success":{"/lights/20/state/bri":254}}]
{"state":{"on":true,"bri":254,"alert":"select","mode":"homeautomation","reachable":false},"swupdate":{"state":"notupdatable","lastinstall":"2020-08-02T03:28:21"},"type":"Dimmable light","name":"Dimmable light 1","modelid":"TRADFRI bulb E27 W opal 1000lm","manufacturername":"IKEA of Sweden","productname":"Dimmable light","capabilities":{"certified":false,"control":{},"streaming":{"renderer":false,"proxy":false}},"config":{"archetype":"classicbulb","function":"functional","direction":"omnidirectional"},"uniqueid":"00:0d:6f:ff:fe:fc:f5:e3-01","swversion":"1.2.214"}
```
As you can see, the state and brightness are reported exactly as what the last command was set; however the reachable attribute is set to false. There's no link between the state and if the light was indeed reachable.

The "reachable" attribute is set independently of the cached value as it's set when creating the Hue json.

So I believe the new behaviour set by this changeset is the correct one, and it gives the closest replicate of the original Hue hub behaviour.

jyavenard replied:
For comparison, here is what the emulated hue will return under the same circumstances:
```
$ curl -X PUT -d '{"on":true,"bri":254}' 192.168.10.5/api/XXX/lights/56/state && echo && curl http://192.168.10.5/api/XXX/lights/56
[{"success": {"/lights/56/state/on": true}}, {"success": {"/lights/56/state/bri": 254}}]
{"manufacturername": "Home Assistant", "modelid": "HASS123", "name": "Kitchen Light 3", "state": {"bri": 254, "mode": "homeautomation", "on": true, "reachable": false}, "swversion": "123", "type": "Dimmable light", "uniqueid": "00:04:db:70:6d:55:14:d8-9d"}
```

As you can see, it's similar to the real Hue, a set and brightness set to the PUT arguments; but a reachable attribute set to false.

bdraco replied:
Right now alexa is reporting failures when the values don't match.  I don't think we should break that functionality.

bdraco reviewed (changes requested):
Please see comments above

jyavenard commented:
If I understand properly you don't want to break a broken functionality :) 

It's the whole point of this change. HA doesn't currently behave as it should. 

Alexa is reporting a failure only with HA because it is indeed reporting a failure that shouldn't be there in the first place.

You never get to see such failure when using a real Hue hub.
So I don't believe we should ever report a failure, despite this is not a failure per say, it's the read back status that don't match the values set. 

Do we want to fully re-implement the Hue protocol and behave differently, or is the aim to emulate a hue hub?

I believe it should be the latter.

jyavenard commented:
I should add that the failure is reflected in the reachable attribute.

bdraco commented:
Caching the PUT values makes sense to solve the precision loss of the brightness.  The concern I have is that we are now reporting an assumed success state for 2 seconds after every PUT request.   If alexa didn't care about the success of the request it doesn't seems like they would have implemented a GET right after the PUT as the PUT would have been sufficient if assume success was the intention here.

jyavenard commented:
> 
> 
> Caching the PUT values makes sense to solve the precision loss of the brightness. The concern I have is that we are now reporting an assumed success state for 2 seconds after every PUT request. If alexa didn't care about the success of the request it doesn't seems like they would have implemented a GET right after the PUT as the PUT would have been sufficient if assume success was the intention here.

What Alexa does doesn't always have to make complete sense in regards to what the Hue actually does (wouldn't be a first for Amazon in my personal experience :) )

From the way I've analysed thing, the GET return the state of the controller, not the state of the entity. Otherwise an unreachable light wouldn't see the brightness or the state reported as being correct.

It seems to me that if a failure is to be returned ; it shouldn't be reported in the GET exchange but in the PUT answer.
E.g. we only reply on the PUT after the command has been actually sent to the entities.

But again, we do report a failure in the sense that reachable is the correct value. And Alexa will show a banner if the device is not reachable.

To summarise, what exactly would you like done to approve the fix for this bug?

We can't report the actual value of the brightness nor the colour setting; they won't match.
That leaves only the value of the state to not be cached.

So would you be okay if only the state value wasn't cached?

bdraco commented:
> 
> So would you be okay if only the state value wasn't cached?

Yes, that is exactly what I am getting at.  If the light fails to turn on, we should report the failure.

jyavenard commented:
> > So would you be okay if only the state value wasn't cached?
> 
> Yes, that is exactly what I am getting at. If the light fails to turn on, we should report the failure.

OK then we need much more logic if we want to do the right thing and remove this 250ms wait.

Not being familiar with the framework, is there an easy way to determine when the action was actually acted upon, so that we only reply to the PUT command once its done.

jyavenard commented:
ok. so bear with me; I've never done any python 3 before today and last I did any real python was a decade ago; so the learning curve isn't minimal.

I have a change that address your concern.
What we will do now is wait for the state change to actually occur before responding to the PUT request.
And for any follow-up GET requests, we will only use the cached values for the attributes other than the state.

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -534,10 +544,35 @@ async def put(self, request, username, entity_number):
+                def async_event_changed(event):
+                    asyncevent.set()
+
+                async def wait_on_change():
+                    await asyncevent.wait()
+
```
`asyncevent.wait()` is already an awaitable so no need to wrap it

bdraco replied:
Suggested change here https://github.com/bdraco/home-assistant/commit/1e4f43e414fd3ab8d5402771d56f8fdacf8112e4

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -534,10 +544,35 @@ async def put(self, request, username, entity_number):
+                async def wait_on_change():
+                    await asyncevent.wait()
+
+                unsub = async_track_state_change_event(
```
The tests mock out the service call so we will need to remove the mocking

bdraco replied:
Suggested change here https://github.com/bdraco/home-assistant/commit/1e4f43e414fd3ab8d5402771d56f8fdacf8112e4

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -534,10 +544,35 @@ async def put(self, request, username, entity_number):
             )
 
+            if state_will_change:
+                # Wait for the state to change.
```
Let's make this its own function so it shows intent

bdraco replied:
Suggested change here https://github.com/bdraco/home-assistant/commit/1e4f43e414fd3ab8d5402771d56f8fdacf8112e4

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -556,16 +591,27 @@ async def put(self, request, username, entity_number):
+        elif time.time() - cached_state_entry[STATE_CACHE_TIME] < STATE_CACHED_TIMEOUT:
+            # We do want to report the actual state of the entity.
+            cached_state = cached_state_entry[STATE_CACHE_STATE]
+            cached_state[STATE_ON] = entity.state != STATE_OFF
```
We should invalidate the cache if its out of sync

bdraco replied:
Suggested change here https://github.com/bdraco/home-assistant/commit/1e4f43e414fd3ab8d5402771d56f8fdacf8112e4

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -556,16 +591,27 @@ async def put(self, request, username, entity_number):
+    cached_state_entry = config.cached_states.get(entity.entity_id, None)
+    cached_state = None
+
+    # Check if we have a cached entry, and if so if it hasn't expired.
+    if cached_state_entry is not None:
+        if cached_state_entry[STATE_CACHE_TIME] is None:
+            # Handle the where the entity is listed in config.off_maps_to_on_domains.
+            cached_state = cached_state_entry[STATE_CACHE_STATE]
+        elif time.time() - cached_state_entry[STATE_CACHE_TIME] < STATE_CACHED_TIMEOUT:
+            # We do want to report the actual state of the entity.
+            cached_state = cached_state_entry[STATE_CACHE_STATE]
+            cached_state[STATE_ON] = entity.state != STATE_OFF
```
Then we need a new test for the caching

jyavenard replied:
Is there more to do than the changes you've already made?

what would you like to test?

bdraco replied:
A test to make sure that the GET response is the same as the PUT response within 2 seconds if the service call is successful and the state changes.

A test to make sure that the GET response does not use the cache if PUT response within 2 seconds if the service call is Unsuccessful and the state does not change.

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -534,10 +544,35 @@ async def put(self, request, username, entity_number):
+                asyncevent = asyncio.Event()
+
+                @core.callback
+                def async_event_changed(event):
```
```suggestion
                def async_event_changed(_):
```

We don't care about this argument so we can use `_`

bdraco replied:
Suggested change here https://github.com/bdraco/home-assistant/commit/1e4f43e414fd3ab8d5402771d56f8fdacf8112e4

bdraco reviewed (changes requested):
This is looking good 👍 .  I linked a sample commit for the suggested cleanups.

After that, we need a test for the caching and this should be good to go.
https://github.com/home-assistant/core/pull/38451/files#r464112980

jyavenard committed:
Add new tests verifying emulated_hue behaviour.

bdraco commented on tests/components/emulated_hue/test_hue_api.py:
```diff
@@ -1042,3 +1060,125 @@ async def test_unauthorized_user_blocked(hue_client):
+    )
+
+    # go through api to get the state back, the value returned should match those set in the last PUT request.
+    ceiling_json = await perform_get_light_state(
```
```suggestion
    with patch() .....
          ceiling_json = await perform_get_light_state(
```

We want to avoid sleeping in tests for more than a brief moment.

patch `STATE_CACHED_TIMEOUT` to be something like `0.00000001` so it times out right away.

jyavenard replied:
If so, I could get intermittent failure for the test before the wait. I can remove it I guess as that functionality is tested just before in a different shape.

bdraco replied:
We probably need two constants ..maybe `STATE_CACHED_TIMEOUT` and `STATE_CHANGE_WAIT_TIMEOUT` so you can patch just the one you want since

jyavenard replied:
I was referring to checking that we get the cache value before the sleep. This one will likely fail with a state_cached_timeout of 1ns.

bdraco replied:
None of the tests are hitting the timeout for the state change event not coming though. 

I'll take a look again in the morning.  It is late late for me and I'm barely awake so I'm probably not so helpful right now.

bdraco commented on tests/components/emulated_hue/test_hue_api.py:
```diff
@@ -1042,3 +1060,125 @@ async def test_unauthorized_user_blocked(hue_client):
+    assert ceiling_json["state"][HUE_API_STATE_SAT] == 127
+    assert ceiling_json["state"][HUE_API_STATE_BRI] == 254
+
+    await asyncio.sleep(STATE_CACHED_TIMEOUT + 0.1)
```
```suggestion
    await asyncio.sleep(0.001)
```

With it patched above we can reduce the sleep

bdraco commented:
You can find the uncovered lines that is making the CI unhappy with

`pytest --cov=homeassistant/components/emulated_hue/ --cov-report term-missing -- tests/components/emulated_hue/`

jyavenard committed:
Increase code test coverage.

The remaining uncovered lines can't be tested as they mostly check that the hass framework or the http server properly work.

This commit doesn't attempt to fix exposed issues as it would be out of scope ; it merely create the tests to exercise the whole code.

jyavenard commented:
> You can find the uncovered lines that is making the CI unhappy with
> 
> `pytest --cov=homeassistant/components/emulated_hue/ --cov-report term-missing -- tests/components/emulated_hue/`

I've added a 3rd commit that cover all codes. Doing the remaining one wouldn't be trivial and I don't think they are necessary: it's mostly ensuring that what hass returns is sane or the http server behaves properly. The only way to get this code covered would be to assert that the condition is always true rather than test for conditions that can't happen.

For the record: this wasn't a fun exercise.

bdraco commented:
We are still missing coverage on 840-841 which is new

 `homeassistant/components/emulated_hue/hue_api.py      412     14    97%   122, 140-141, 183-186, 300-301, 339-340, 498, 590, 840-841`

I'll see if I can write a test for that later today.

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -66,10 +67,17 @@
 
+# How long an entry state's cache will be valid for in seconds.
+STATE_CACHED_TIMEOUT = 2.0
+# Index value when accessing a cached state.
+STATE_CACHE_STATE = 0
+STATE_CACHE_TIME = 1
```
```suggestion
```

We can get rid of this now that is been refactored

bdraco committed:
Update homeassistant/components/emulated_hue/hue_api.py

bdraco committed:
Add test for state change wait timeout

bdraco commented:
I added the coverage for the two lines that were missing.  The CI should pass now.

bdraco commented:
Testing checks out for me.  

@jyavenard Is the 2s timeout long enough for the group to change state or should we increase that?

bdraco commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -520,7 +527,9 @@ async def put(self, request, username, entity_number):
             # they'll map to "on". Thus, instead of reporting its actual
             # status, we report what Alexa will want to see, which is the same
             # as the actual requested command.
-            config.cached_states[entity_id] = parsed
+            config.cached_states[entity_id] = [parsed, None]
+        else:
+            config.cached_states[entity_id] = [parsed, time.time()]
```
We should probably move this code (just the part about setting the cache) after the `wait_for_state_change_or_timeout` so the cache lasts for the full 2 seconds after the successful state change.

bdraco replied:
That should better handle your group case.

bdraco replied:
I did some testing with groups, and it is not longer enough.  I made the adjustment.

bdraco committed:
Preserve the cache long enough for groups to change

bdraco reviewed (approved):
Retested, and everything appears to be working with the `group` case described above.

Please confirm.

jyavenard commented:
> 
> 
> Testing checks out for me.
> 
> @jyavenard Is the 2s timeout long enough for the group to change state or should we increase that?

I've had some errors when using my "All Lights" group which is made of about 40 zigbee devices.
I don't know if using more than 2s would help, however being a timeout and seeing that we only wait "up to" that amount I don't think that using a longer value will cause issue.


Should we take the 3rd patch and move it to its own bug?

bdraco commented:
> > Testing checks out for me.
> > @jyavenard Is the 2s timeout long enough for the group to change state or should we increase that?
> 
> I've had some errors when using my "All Lights" group which is made of about 40 zigbee devices.
> I don't know if using more than 2s would help, however being a timeout and seeing that we only wait "up to" that amount I don't think that using a longer value will cause issue.

How high do we need to take the timeout for your light group to work as expected?

> 
> Should we take the 3rd patch and move it to its own bug?

jyavenard commented on tests/components/emulated_hue/test_hue_api.py:
```diff
@@ -1310,6 +1310,57 @@ async def test_put_then_get_cached_properly(hass, hass_hue, hue_client):
+        )
+
+    # Ensure we did not actually turn on
+    assert hass.states.get("light.ceiling_lights").state == STATE_OFF
```
The first await serves no purpose, it won't affect the put_light_state method as it starts a new timer.

I think this could intermittently fail as there's no guarantee the service won't actually be on (it's all async) with no wait.

bdraco replied:
```
    hass_hue.services.async_register(
        light.DOMAIN, SERVICE_TURN_ON, mock_service_call, schema=None
    )
```

This is mocking out the turn on service call so it is not possible to turn the light on with the service anymore.

jyavenard replied:
Oh I see.

I don't see how the await line 1340 makes a difference to anything though.

bdraco replied:
It shouldn't be needed.

jyavenard commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -73,7 +73,7 @@
 
 # How long to wait for a state change to happen
-STATE_CHANGE_WAIT_TIMEOUT = 2.0
+STATE_CHANGE_WAIT_TIMEOUT = 5.0
```
I can make some local test with Alexa to see how long it will wait for before aborting and giving an error.

Then we can make the timeout value just slightly below that.

jyavenard commented on homeassistant/components/emulated_hue/hue_api.py:
```diff
@@ -72,6 +72,8 @@
 _LOGGER = logging.getLogger(__name__)
 
+# How long to wait for a state change to happen
+STATE_CHANGE_WAIT_TIMEOUT = 2.0
```
I'm not sure having separated value makes much difference nor if it's something we want.

One is how long we wait for the state to change (and it may not always change like if we're just changing the contrast). The first one guarantees that the group/light did turn on/off.

The other make sure we trick Alexa long enough to believe we did change things properly as changing attributes like brightness somehow take much longer than changing the state.

The wait for the change is to potentially report an error; but my experience with a real Hue hub is that an error is never reported. Like you're in a room where the power to a light is off: you say Alexa turn the light on and she will say "OK" but nothing happens. Only in the app will you see a banner that the light isn't responding.

As far as Alexa is concerned; only the STATE_CACHED_TIMEOUT value matters, to ensure the GET that follows the PUT gets the same value

jyavenard reviewed (commented):
indeed.

Should have seen that issue; the cache's time should only be set after we've replied to the PUT request.

I think this is the real change needed here.

bdraco commented:
- [ ] Retest with desktop app
- [ ] Retest with harmony

bdraco commented:
Verified everything is working with harmony and colors for hue desktop on macos

bdraco commented on tests/components/emulated_hue/test_hue_api.py:
```diff
@@ -1042,3 +1185,292 @@ async def test_unauthorized_user_blocked(hue_client):
+    assert ceiling_lights.state == STATE_OFF
+
+    with patch.object(hue_api, "STATE_CHANGE_WAIT_TIMEOUT", 0.000001):
+        await asyncio.sleep(0.000001)
```
```suggestion
```

bdraco committed:
Update tests/components/emulated_hue/test_hue_api.py
